editor/src/clj/internal/node.clj

```diff
@@ -1334,11 +1334,14 @@
     (type-name (gt/node-type node))))
 
 (defn mark-in-production [node-id label evaluation-context]
-  (assert (not (contains? (:in-production evaluation-context) (gt/endpoint node-id label)))
-          (format "Cycle detected on node type %s and output %s"
-                  (node-type-name node-id evaluation-context)
-                  label))
-  (update evaluation-context :in-production conj (gt/endpoint node-id label)))
+  (if (contains? (:in-production evaluation-context) (gt/endpoint node-id label))
+    (throw (ex-info (format "Cycle detected on node type %s and output %s"
+                      (node-type-name node-id evaluation-context)
+                      label)
+                    {:cause "cycle-detected"
```

Comment:
These kind of "comparable strings" that are typically used as an enum are preferably expressed as [keywords](https://clojure.org/reference/data_structures#Keywords). Keywords are readable as strings but their strings are interned in a shared storage and comparisons are extremely cheap as a result.
```suggestion
                    {:cause :cycle-detected
```